### ISSUE TITLE
ForEach-Object -Parallel disallows script blocks via $using:, unlike Start-ThreadJob

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#12378]
`Start-ThreadJob` allows passing a script block by way of a `$using:` reference, which is a convenient way to make a function from the caller's scope available to the job:

```powershell
PS> function foo { 'bar' }; Start-ThreadJob { & $using:function:foo } | Receive-Job -Wait -AutoRemoveJob
bar
```

That is, the `foo` function's script block (`[scriptblock]` instance containing the function body) reported by `$function:foo` (namespace variable notation) was referenced via `$using:`, allowing it to be called with `&`.

Unexpectedly, `ForEach-Object -Parallel`, whose behavior is generally very similar to `Start-ThreadJob`, _explicitly disallows_ referencing `[scriptblock]`-typed values from the caller's scope via `$using:`

Note:

* The error message (see below) mentions "undefined behavior" - but wouldn't that apply to `Start-ThreadJob` as well? Is the answer to disallow script blocks in _both_?
  
  * If the concern is thread safety: that problem exists for all reference types passed via `$using:`, not just for script blocks that may form closures with outside values from the caller's scope.

* #12240  proposes an opt-in mechanism for `ForEach-Object -Parallel` to allow copying the calling runspace's state to the parallel threads, which would allow for direct use of function from the caller's scope; however, a lower-overhead way of being able to call functions _selectively_ still seems beneficial.

* Loosely related: #11698

# Steps to reproduce

```powershell
function foo { 'bar' }; ForEach-Object -Parallel { & $using:function:foo } | Should -Be 'bar'
```

# Expected behavior

The test should succeed.

# Actual behavior

The test fails, because a statement-terminating error occurs:

```none
A ForEach-Object -Parallel using variable cannot be a script block. 
Passed-in script block variables are not supported with ForEach-Object -Parallel, 
and can result in undefined behavior.
```

# Environment data

```none
PowerShell Core 7.1.0-preview.1
```

**Comment by bpayette** [comment]
@mklement0  ScriptBlocks are bound to the runspace where they were created so you can't reliably invoke them in a different runspace without regenerating the scriptblock  in the new runspace. Fortunately regenerating the script block is pretty cheap using the existing (internal) APIs. I'm guessing @PaulHigin is doing this for thread jobs but not for `foreach -parallel`. Paul?

BTW - another thing to check out are classes. An instance of a PowerShell class is also affinitized to the runspace where it was created, so passing that instance to another runspace won't work reliably.

**Comment by PaulHigin** [comment]
As @bpayette states, the concern was script block affinity to the runspace in which it was created.  As @mklement0  mentions, this should also be disallowed for Start-ThreadJob.  And I agree with @mklement0 that with the new upcoming optional "replicate runspace state" feature, it may make sense to allow a scriptblock variable by re-creating it in the new runspace.  I'll include that as part of the feature.

Thanks!

**Comment by vexx32** [comment]
Users consistently seem to come to these kinds of features wanting to be able to transfer a scriptblock to the new runspace or process and have it run. This isn't possible to do directly due to those concerns, but perhaps there's  something we can do around automatically serializing/deserializing the scriptblock as it crosses the runspace boundary when the session state isn't replicated.

Even if it's an opt in via a new `[SerializableScriptBlock]` sort of type that automatically re-hydrates the scriptblock and binds itself to the new runspace it finds itself in, I think users will find that experience much more intuitive than simply preventing scriptblocks to be passed around at all.

**Comment by PaulHigin** [comment]
I don't think it makes sense to re-create a new scriptblock unless the runspace in which it is being created is replicated from the client script block, because it (the scriptblock) may be referencing variables, functions, modules that don't exist.  But I see no reason not to do it into a replicated runspace, other than known variable mulit-threading issues.

/cc @daxian-dbw

**Comment by vexx32** [comment]
It _might_ be, sure. But just as often as I see that, I'll see someone trying to throw the same self-contained script into a handful of different threads to parallelize a task.

For a case like that, I fail to see why permitting such a thing meets so much resistance. 😕

**Comment by mklement0** [comment]
I agree with @vexx32: simply transparently recreating the script block in the target runspace - whether it is a replicated one or not -  is sensible default behavior that will cover a lot of use cases, notably the case where the script block is a _function body_ or a similarly _self-contained_ script block, as in the example in the OP.

All that is then needed is to _document_ the behavior and limitations, in the context of the `$using:` specifier  in the `about_Remote_Variables` topic.

I also think it is the sensible default behavior in remoting / background job scenarios: see #11698

**Comment by bpayette** [comment]
@PaulHigin IIRC  you can quickly create a new script block by using the AST of the existing script block. There is a non-public API (or constructor?) on scriptblock to do this. I measured the performance and it was quit good. This will be especially important if you're implementing a `Runspce.Clone()`, as you'll have to do this for every function being copied to the new runspace.

**Comment by vexx32** [comment]
Not sure about scriptblock itself, but ScriptblockAst has a `GetScriptBlock()` method, at least. 🙂

**Comment by Viajaz** [comment]
Maybe a different approach but perhaps `ForEach-Object -Parallel` with `-AsThreadJob` similar to that described in https://github.com/PowerShell/PowerShell/issues/10841 ?

https://github.com/MicrosoftDocs/PowerShell-Docs/issues/4977 implies that `ForEach-Object -Parallel -AsJob` is a ThreadJob but it's a [PSTaskJob](https://github.com/PowerShell/PowerShell/issues/10841#issuecomment-545193567) isn't it?

Docs imply something similar: https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_thread_jobs?view=powershell-7.1#thread-jobs-and-variables

**Comment by segraef** [comment]
Hi there, #16564 fixed #16445 but is it part of [v7.3.0-preview.1](https://github.com/PowerShell/PowerShell/compare/v7.2.0-preview.10...v7.3.0-preview.1) or [7.2.1](https://github.com/PowerShell/PowerShell/releases/tag/v7.2.1)?

**Comment by kasini3000** [comment]
hi who add backport label？
from i tested，ps 7.2 ps6.2 has this problem.
but ps 7.0,ps7.1 normal
